shared/src/main/scala/effekt/symbols/Name.scala

```diff
@@ -4,26 +4,117 @@ import effekt.source.Id
 import effekt.context.Context
 
 trait Name {
-  def name: String
+  /**
+   * The local part of the name relative to its potential parent.
+   *
+   * @example The local part of the name `"foo.baz.bar"` is `"bar"`.
+   */
+  val localName: String
 
-  // this should only be used for reporting errors, not for code generation
-  def qualified: String
-  override def toString = name
-  def rename(f: String => String): Name
-}
+  /**
+   * Computes the qualified version of this name.
+   * @return A String containing the local part of this name prefixed by their parent name. Components are separated by the point character (`.`).
+   *
+   * @note this should only be used for reporting errors, not for code generation.
+   * @note Note to the previous note: Is this still true for the new module-system?
+   */
+  def qualifiedName: String = this match {
+    case ToplevelName(name)       => name
+    case NestedName(parent, name) => s"${parent.qualifiedName}.$localName"
+  }
```

Comment:
I would say it is typically better style (at least more consistent with the remaining code base) to use overwrite `qualifiedName` in the subclass, then pattern matching on `this`.